### DISCUSSION TITLE
Extracting T function type

### DISCUSSION THREAD

**Discussion by ScreamZ** [amannn/next-intl#1331] [Q&A] [answered]
Hi @amannn,

I'm trying to execute the https://next-safe-action.dev/docs/recipes/i18n recipes, for that I need to access the T function type to get proper DX.

> [!NOTE]
> After some experimentation, I have succeeded in extracting the typings correctly. Maybe we could add this to the documentation.

```ts
type TFn =
  | ReturnType<typeof useTranslations<never>>
  | Awaited<ReturnType<typeof getTranslations<never>>>;

export const authSchema = (t: TFn) => {
  return z.object({
    username: z.string(),
    password: z.string(),
    resellerCode: z.coerce.number(),
  });
};
```

```ts
import { getTranslations } from "next-intl/server";

export async function MyPage  {
   const t = await getTranslations()
  const schema = authSchema()
}
```

```tsx
export const AuthenticationForm = () => {
  const t = useTranslations();
  const schema = authSchema(t);
}
```

**Comment by ScreamZ** [comment]
@amannn Any thoughts on adding that to the documentation? I might submit a pull request

**Reply by amannn** [reply, marked as answer]
Hey @ScreamZ!

It's really kind of you to offer a pull request, but as discussed in https://github.com/amannn/next-intl/discussions/1134, the snippet that's shown in the `next-safe-action` docs is unfortunately not what I'd recommend.

There's now a basic [`zod` example in the docs](https://next-intl-docs.vercel.app/docs/environments/actions-metadata-route-handlers#server-actions-zod). As for `next-safe-action`, I think [trying out `handleValidationErrorsShape`](https://github.com/amannn/next-intl/discussions/1134#discussioncomment-9840159) could be something to look into (and if this works, maybe a PR to the `next-safe-action` docs could be helpful).

**Reply by ScreamZ** [reply]
Yes, I understand.

Currently I'm using this solution. But I agree this could be better using https://next-safe-action.dev/docs/define-actions/validation-errors as we already discussed in the other topic. I need to spend time trying it. I'll push a card onto my trello for next weeks.

Thanks
